openedx/core/djangoapps/monitoring_utils/__init__.py

```diff
@@ -0,0 +1,97 @@
+"""
+This is an interface to the monitoring_utils middleware.  Functions
+defined in this module can be used to report monitoring custom metrics.
+
+Usage:
+
+    from openedx.core.djangoapps import monitoring_utils
+    ...
+    monitoring_utils.accumulate('xb_user_state.get_many.num_items', 4)
+
+There is no need to do anything else.  The metrics are automatically cleared
+before the next request.
+
+We try to keep track of our custom metrics at:
+
+https://openedx.atlassian.net/wiki/display/PERF/Custom+Metrics+in+New+Relic
+
+At this time, these custom metrics will only be reported to New Relic.
+
+TODO: supply additional public functions for storing strings and booleans.
+
+"""
+
+from . import middleware
+try:
+    import newrelic.agent
+except ImportError:
+    newrelic = None  # pylint: disable=invalid-name
+
+
+def accumulate(name, value):
+    """
+    Accumulate monitoring custom metric for the current request.
+
+    The named metric is accumulated by a numerical amount using the sum.  All
+    metrics are queued up in the request_cache for this request.  At the end of
+    the request, the monitoring_utils middleware will batch report all
+    queued accumulated metrics to the monitoring tool (e.g. New Relic).
+
+    Arguments:
+        name (str): The metric name.  It should be period-delimited, and
+            increase in specificity from left to right.  For example:
+            'xb_user_state.get_many.num_items'.
+        value (number):  The amount to accumulate into the named metric.  When
+            accumulate() is called multiple times for a given metric name
+            during a request, the sum of the values for each call is reported
+            for that metric.  For metrics which don't make sense to accumulate,
+            make sure to only call this function once during a request.
+    """
+    middleware.MonitoringCustomMetrics.accumulate_metric(name, value)
+
+
+def increment(name):
+    """
+    Increment a monitoring custom metric representing a counter.
+
+    Here we simply accumulate a new custom metric with a value of 1, and the
+    middleware should automatically aggregate this metric.
+    """
+    accumulate(name, 1)
+
+
+def add_custom_metrics_for_course_key(course_key):
```

Comment:
Do you think we could rename this from `add_` to `set_`?  Now that these are alongside the `aggregate` and `increment` functions which have "memory", I would like to emphasize that these functions do *not* have memory by calling them `set`ters.